README.md
<p align="center">
    <a href="http://www.imac-project.eu/">
        <img height="300px" width="774px" src="./portal/img/LOGO-IMAC.png" />
    </a>
</p>

## Getting Started

These instructions will get you a copy of the project up and running on your local machine for development and testing purposes. See deployment for notes on how to deploy the project on a live system.

## Features

The ImAc video has a lits of editable access services with multi options and features:

* [Subtitles](https://github.com/ua-i2cat/ImAc/wiki/Subtitles)
* [Sign Language](https://github.com/ua-i2cat/ImAc/wiki/Sign-Language)
* [Audio Description](https://github.com/ua-i2cat/ImAc/wiki/Audio-Description)
* [Audio Subtitles](https://github.com/ua-i2cat/ImAc/wiki/Audio-Subtitles)


### Prerequisites

What things you need to install the software and how to install them

<ul>
    <li> Apache Tomcat (or any other web server)</li>
    <li> Git </li>
</ul>


### Installing

How to use run the ImAc player in Apache Tomcat:

* Run C:\apache-tomcat\bin\startup.bat
* Open C:\apache-tomcat\webapps folder
* Copy imac_player folder into webapps folder
* Open a browser and type: "your_apache_server_address"/imac_player/


## Built With

* [IMSC_360.js](https://github.com/sandflow/imscJS) - JavaScript library for rendering IMSC Text and Image Profile documents in HTML5
* [three.js](https://threejs.org/)
* [DASH.js](https://github.com/Dash-Industry-Forum/dash.js/wiki) - Reference client implementation for the playback of DASH contents via JavaScript and compliant browsers.
* [Omnitone.js](https://github.com/GoogleChrome/omnitone) - Implementation of Ambisonic decoding and binaural rendering written in Web Audio Application Program Interface


## Acknowledgments

This project has received funding from the European Union’s Horizon 2020 research and innovation programme under grant agreement [No 761974](https://cordis.europa.eu/project/rcn/211084/factsheet/en).


CHANGELOG.md
# Release 0.8.0

Date 27/06/2019

## Changelog

	* Add: Change SL size
	* Add: Holy land 4 description
	* Add: New covers
	* Add: Only access languages with available languages are shown
	* Add: New banner images
	* Add: Save config into the player menu
	* Add: remote bootstrap libs
	* Add: Google Analytics
    * Add: unique id per device
	* Modify: Acces options are disabled when language is not available
	* Modify: Open links in a new tap
	* Modify: New MainMenuController that includes the old playpause volume settings and access controllers
	* Modify: New MainMenuModel that includes the old playpause volume settings and access models
	* Modify: Added videoProgressBarController to MainMenuController
	* Fix: Issue with SL arrows
	* Fix: Error in presentation title
	* Fix: Presentation AD/AST listing when available
	* Fix: Bug when active access option was unavailable and then available again could not be clicked.
	* Fix: German translation
	* Fix: Preview auto play when video was already pasued fixed
	* Fix: When subtitles are disabled signer goes back to bottom
	* Fix: Issue with search check box
	* Fix: Issues with voice control
	* Fix: Issues with data variable in mainmenucontroller
	* Fix: Options menu title not hidden on start
	* Fix: uuid issue


### Known Bugs/Issues

	* Flickering in the subtitles when move the head in Oculus Go and Samsung VR devices
	* When small menu open subtitles and signer elements are behind


# Release 0.7.0

Date 9/05/2019

## Changelog

	* Add: New portal interface
	* Add: Search functionality in portal interface
	* Add: Translate functionality in portal interface
	* Add: Configuration menu in portal interface
	* Add: Voice control functionalities
	* Add: Buttons to enable or dissable the access. services (ST , SL , AD, AST) in the main menu player interface
	* Add: Representation mode in AST services
	* Add: Disable access options when language not available
	* Add: Change SL size option
	* Add: Links to the parthners webpages and ImAc social networks 	
	* Modify: Player menu interface
	* Modify: Configuration menu structure
	* Modify: content.json adding duration, acces, description, poster and language parameters
	* Modify: Access language, now there is only one language selector for all services
	* Modify: Indicator, now there is the same for ST and SL
	* Modify: Safe area, now there is the same for ST and SL
	* Fix: Latitude subtitles position that is printed inverted
	* Fix: Issue with SL arrows indicator
	* Fix: Vocabulary issues

### Known Bugs/Issues

	* Flickering in the subtitles when move the head in Oculus Go and Samsung VR devices
	* When small menu open subtitles and signer elements are behind


	
	
	
# Release 0.5.0

Date 28/11/2018

## Changelog

	* Add: AD functionalities (language, presentation mode, volume)
	* Add: AST functionalities (language, easy to read, volume)
	* Add: new KPI to the QoE logger
	* Add: support to new MPD with subtitles AdaptationSet
	* Add: support to new MPD with signer AdaptationSet
	* Add: support to new MPD with AD AdaptationSet
	* Add: support to new MPD with AST AdaptationSet
	* Add: support to multiple period in the MPD
	* Add: preview function
	* Add: close multioptions menu when click outside	
	* Modify: The SL Top/Bottom position options has been modified for Right/Left position options
	* Modify: In HMD the menu is opened in front of the user
	* Modify: positioning signer top when subtitles go top
	* Fix: error when the JSON file have an empty signer value
	* Fix: autoplay of the signer video when the Enhanced-Accessibility menu is opened
	* Fix: default options of SL not highlighted in yellow
	* Fix: values for signer area
	* Fix: issues with the oculus rift controller

### Known Bugs/Issues

	* Flickering in the subtitles when move the head in Oculus Go and Samsung VR devices
	* Auto-positioning functionality does not let you open the menu again
	* When small menu open subtitles and signer elements are behind




js/preTest.js

var myVar;
var timerCounter = 0; // per al pretest

function runDemo() 
{
  console.log("Running demo " + demoId)

////////////////
//   Demo 1   //
////////////////
  if (demoId == 0)
  {
    var imgUrl = language == "catala" ? './resources/preTest/cat/1.png' : './resources/preTest/ger/1.png';
    moData.createImageInCamera( imgUrl, 'pauseimg', { w: 1, h: 1, visible: true } );

    myVar = setTimeout(function() 
    {  
      showInfoImages( 2 );

      timerCounter = 0;

      var listVideoContent = moData.getListOfVideoContents();
      var _videoElement = listVideoContent[0].vid;
      AudioManager.initializeAudio( _videoElement, 1, camera.matrixWorld.elements );

      isSubtitleEnabled = true;
      viewArea = 30;
      forcedDisplayAlign = 'after';
      forcedTextAlign = 'center';

      //initSubtitle( viewArea, forcedDisplayAlign, forcedTextAlign, subtileIndicator );

    }, 3000);
  }

////////////////
//   Demo 2   //
////////////////
  else if (demoId == 2)
  {
    var imgUrl = language == "catala" ? './resources/preTest/cat/21.png' : './resources/preTest/ger/21.png';
    moData.createImageInCamera( imgUrl, 'pauseimg', { w: 1, h: 1, visible: true } );

    myVar = setTimeout(function() 
    {  
      showInfoImages( 22 );

      timerCounter = 0;

      var listVideoContent = moData.getListOfVideoContents();
      var _videoElement = listVideoContent[0].vid;
      AudioManager.initializeAudio( _videoElement, 1, camera.matrixWorld.elements );

      subtitleIndicator = 'arrow';
      isSubtitleEnabled = true;
      viewArea = 70;
      forcedDisplayAlign = 'after';
      forcedTextAlign = 'center';

      //initSubtitle( viewArea, forcedDisplayAlign, forcedTextAlign, subtileIndicator );
    }, 3000);

    //runDemo2(30000);
  }

////////////////
//   Demo 3   //
////////////////
  else if (demoId == 3)
  {
    var imgUrl = language == "catala" ? './resources/preTest/cat/32.png' : './resources/preTest/ger/32.png';
    moData.createImageInCamera( imgUrl, 'pauseimg', { w: 1, h: 1, visible: true } );

    myVar = setTimeout(function() 
    {  
      showInfoImages( 33 );

      timerCounter = 0;

      var listVideoContent = moData.getListOfVideoContents();
      var _videoElement = listVideoContent[0].vid;
      AudioManager.initializeAudio( _videoElement, 1, camera.matrixWorld.elements );

      //initSign( viewArea, signArea, signIndicator )

      var signPosition = getPlanePosition();
      var conf = {
        size: 0.3,
        x: signPosition.x,
        y: signPosition.y,
        z: signPosition.z
      };
      moData.createSignVideo('./resources/signer_rbb_new.mp4', 'mp4', 'sign', conf);

      viewArea = 30;
    }, 3000);

    //runDemo3(30000);
  }

////////////////
//   Demo 4   //
////////////////
  else if (demoId == 4)
  {
    var imgUrl = language == "catala" ? './resources/preTest/cat/52.png' : './resources/preTest/ger/52.png';
    moData.createImageInCamera( imgUrl, 'pauseimg', { w: 1, h: 1, visible: true } );

    myVar = setTimeout(function() 
    {  
      showInfoImages( 53 );

      timerCounter = 0;

      var listVideoContent = moData.getListOfVideoContents();
      var _videoElement = listVideoContent[0].vid;
      AudioManager.initializeAudio( _videoElement, 1, camera.matrixWorld.elements );

      viewArea = 70;
      signIndicator = 'arrow';

      //initSign( viewArea, signArea, signIndicator )

      var signPosition = getPlanePosition();
      var conf = {
        size: 0.3,
        x: signPosition.x,
        y: signPosition.y,
        z: signPosition.z
      };
      moData.createSignVideo('./resources/signer_rbb_new.mp4', 'mp4', 'sign', conf);

    }, 3000);

    //runDemo4(30000);
  }

////////////////
//   Demo 5   //
////////////////
  else if (demoId == 5)
  {
    var imgUrl = language == "catala" ? './resources/preTest/cat/60.png' : './resources/preTest/ger/60.png';
    moData.createImageInCamera( imgUrl, 'pauseimg', { w: 1, h: 1, visible: true } );

    myVar = setTimeout(function() 
    {  
      showInfoImages( 61 );

      timerCounter = 0;
      
      var listVideoContent = moData.getListOfVideoContents();
      var _videoElement = listVideoContent[0].vid;
      AudioManager.initializeAudio( _videoElement, 1, camera.matrixWorld.elements );

      viewArea = 70;
      autoPositioning = 'enable';

      //initSign( viewArea, signArea, signIndicator )
      //enableAutoPositioning()

      var signPosition = getPlanePosition();
      var conf = {
        size: 0.3,
        x: signPosition.x,
        y: signPosition.y,
        z: signPosition.z
      };
      moData.createSignVideo('./resources/signer_rbb_new.mp4', 'mp4', 'sign', conf);

    }, 3000);

  }

////////////////
//   Demo 6   //
////////////////
  else
  {
    var listVideoContent = moData.getListOfVideoContents();
    var _videoElement = listVideoContent[0].vid;
    AudioManager.initializeAudio( _videoElement, 2, camera.matrixWorld.elements );

    var controlBar = moData.createControlBar();
    console.log(controlBar);
    interController.addInteractiveObject(controlBar);

    subController.enableSubtitles();
    subController.initSubtitle( 60, 0, -1, 'arrow' );
    subController.setSubtitle( "./resources/Rapzember_Cat.xml" );  

    //subController.enableAutoPositioning();

    moData.playAll();
  }
}


function showInfoImages(id1)
{
  moData.pauseAll();
  var imgUrl = language == "catala" ? './resources/preTest/cat/' + id1 + '.png' : './resources/preTest/ger/' + id1 + '.png';
  moData.createImageInCamera( imgUrl, 'pauseimg', { w: 1, h: 1, visible: true } );

  myVar = setTimeout(function() 
  {
    moData.removeInfoImage();

    if (demoId == 1) runDemo1( 10000 );
    else if (demoId == 2) runDemo2( 30000 );
    else if (demoId == 3) runDemo3( 30000 );
    else if (demoId == 4) runDemo4( 30000 );
    else if (demoId == 5) runDemo5( 60000 );

  }, 3000);
}

function showEndImage(id1)
{
  moData.pauseAll();
  var imgUrl = language == "catala" ? './resources/preTest/cat/' + id1 + '.png' : './resources/preTest/ger/' + id1 + '.png';
  moData.createImageInCamera( imgUrl, 'pauseimg', { w: 0.8, h: 0.8, visible: true } );

  myVar = setTimeout(function() 
  {
    location.reload();
  }, 3000);
}

function waitResponse(id)
{
  if ( moData.isPausedById(0) ) 
  {
    var imgUrl = language == "catala" ? './resources/preTest/cat/' + id + '.png' : './resources/preTest/ger/' + id + '.png';
    moData.createImageInCamera( imgUrl, 'pauseimg', { w: 1, h: 1, visible: true } );

    myVar2 = setInterval(function() {
      timerCounter++;
      if(timerCounter == 20) {
        moData.playAll();
      }
      if (!moData.isPausedById(0)) 
      {
        clearInterval(myVar2);
        moData.removeInfoImage(); 
        if (demoId == 1) 
        {
          if (viewArea == 40) showInfoImages( 5 );
          else if (viewArea == 50) showInfoImages( 8 );
          else if (viewArea == 60) showInfoImages( 11 );
          else if (viewArea == 70) showInfoImages( 14 );
          else if (viewArea == 80) showInfoImages( 17 );
          else showEndImage( 20 );
        }   
        else if(demoId == 2)
        {
          if (subtitleIndicator == 'compass') showInfoImages( 25 );
          else if (subtitleIndicator == 'move') showInfoImages( 28 );
          else showEndImage( 31 );
        }  
        else if (demoId == 3) 
        {
          if (viewArea == 40) showInfoImages( 36 );
          else if (viewArea == 50) showInfoImages( 39 );
          else if (viewArea == 60) showInfoImages( 42 );
          else if (viewArea == 70) showInfoImages( 45 );
          else if (viewArea == 80) showInfoImages( 48 );
          else showEndImage( 51 );
        } 
        else if(demoId == 4)
        {
          if (signIndicator == 'move') 
          {
            showInfoImages( 56 );
            var signPosition = getPlanePosition();
            var conf = {
              size: 0.3,
              x: signPosition.x,
              y: signPosition.y,
              z: signPosition.z
            };
            moData.createSignVideo('./resources/signer_rbb_new.mp4', 'mp4', 'sign', conf);
          }
          else showEndImage( 59 );
        }  
        else if(demoId == 5)
        {
          showEndImage( 64 );
        }
      }
    }, 500);
  }
}

function runDemo1(time) 
{ 
  moData.playAll();
  myVar = setTimeout(function() 
  {
    timerCounter = 0;
    if (forcedTextAlign == 'center') 
    {
      if (forcedDisplayAlign == 'before') 
      {         
        moData.pauseAll();
        forcedDisplayAlign = 'after';
        viewArea += 10;
        textListMemory = [];
        var idDiapo = 4 + 3*(viewArea-40)/10;
        waitResponse(idDiapo);
      } 
      else 
      {
        forcedTextAlign = 'start';
        textListMemory = [];
        runDemo1(10000);
      }
    }
    else 
    {
      forcedDisplayAlign = 'before';
      forcedTextAlign = 'center';
      textListMemory = [];
      runDemo1(10000);
    }
  }, time);  
}

function runDemo2(time) 
{
  moData.playAll();
  myVar = setTimeout(function() 
  {
    timerCounter = 0;
    moData.pauseAll();
    if (subtitleIndicator == 'arrow') 
    {
      subtitleIndicator = 'compass';
      textListMemory = [];
      waitResponse(24);
    }
    else if (subtitleIndicator == 'compass')
    {
      subtitleIndicator = 'move';
      textListMemory = [];
      waitResponse(27);
    }
    else if (subtitleIndicator == 'move')
    {
      subtitleIndicator = 'none';
      textListMemory = [];
      waitResponse(30);
    }
    else 
    {
      location.reload();
    }
  }, time); 
}

function runDemo3(time) 
{
  moData.playAll();
  myVar = setTimeout(function() 
  {
    timerCounter = 0;
    moData.pauseAll();
    moData.removeSignVideo();
    viewArea += 10;
    var signPosition = getPlanePosition();
	  var conf = {
	    size: 0.3,
	    x: signPosition.x,
	    y: signPosition.y,
	    z: signPosition.z
	  };
	  moData.createSignVideo('./resources/signer_rbb_new.mp4', 'mp4', 'sign', conf);
    waitResponse(35);
  }, time); 
}


function runDemo4(time) 
{
  moData.playAll();
  myVar = setTimeout(function() 
  {
    timerCounter = 0;
    moData.pauseAll();
    moData.removeSignVideo();
    if (signIndicator == 'arrow') 
    {
      signIndicator = 'move';
      waitResponse(55);;
    }
    else if (signIndicator == 'move')
    {
      signIndicator = 'none';
      waitResponse(58);
    }
    else {
      location.reload();
    }
  }, time); 
}

function runDemo5(time) 
{
  moData.playAll();
  myVar = setTimeout(function() 
  {
    timerCounter = 0;
    moData.pauseAll();
    moData.removeSignVideo();
    waitResponse(63);

  }, time); 
}


function runADDemo(id)
{
  enterfullscreen();
  //showLoader();
  //myVar = setTimeout(function() {
    //clearLoader();
    //enterfullscreen();
    document.getElementById("ad1").style.display = "block";
    if (id == 0) document.getElementById("ad2").src = "https://www.youtube.com/embed/lyd4HDUA4JY?autoplay=1";//window.location = 'https://youtu.be/SqIrXu0NkVU';
    else if (id == 1) document.getElementById("ad2").src = "https://www.youtube.com/embed/KSyvf5R0Se0?autoplay=1";//window.location = 'https://youtu.be/cMeLWQsU27Y';
    else if (id == 2) document.getElementById("ad2").src = "https://www.youtube.com/embed/akfVIe5NGzk?autoplay=1";//window.location = 'https://youtu.be/x0HAOLUjn1Y';

  //}, 10000); 
}

